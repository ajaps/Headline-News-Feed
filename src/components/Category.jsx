import classNames from 'classnames';
import React from 'react';
import PropTypes from 'prop-types';

import * as myActions from '../actions/Headlines';
import SourcesStore from '../stores/Sources';

/**
 * Represents a navigation bar containing different categories
 */
export default class CategoryComponent extends React.Component {
  constructor() {
    super();
    this.setCategory = this.setCategory.bind(this);
  }

/**
 * sends an action to change category based on user selection
 * @param {object} e The selected category
 * @return {void}
 */
  setCategory(e) {
    const selectedCategory = e.target.dataset.category;
    const url = SourcesStore.getSourceUrl;
    myActions.setCategory(url, selectedCategory);
  }

  render() {
    const categoryObj = {};
    const currentCategory = this.props.currentcategory[0];
    const allSources = this.props.allCategory;
    const filterCategoryInSource = allSources.map((categoryItem) => {
      const eachCategory = categoryItem.category;
      const btnClass = classNames({
        active: currentCategory === eachCategory,
      });
      if (!(eachCategory in categoryObj)) {
        categoryObj[eachCategory] = eachCategory;
        return (
          <li className={btnClass} key={eachCategory} >
            <a data-category={eachCategory} onClick={this.setCategory}
                ref={eachCategory}
            >{eachCategory.split('-')[0]}</a>
          </li>
        );
      }
      return null;
    });

    return (
      <nav className="navbar navbar-default categoryList">
        <div className="container-fluid">
          <div className="navbar-header">
            <button aria-controls="navbar" aria-expanded="false"
                className="navbar-toggle collapsed" data-target="#navbar"
                data-toggle="collapse" type="button"
            >
              <span className="sr-only">Toggle navigation</span>
              <span className="icon-bar" />
              <span className="icon-bar" />
              <span className="icon-bar" />
            </button>
          </div>
          <div className="navbar-collapse collapse" id="navbar" >
            <ul className="nav nav-justified">
              { filterCategoryInSource }
            </ul>
          </div>
        </div>
      </nav>
    );
  }
}

CategoryComponent.propTypes = {
  allCategory: PropTypes.array,
  currentcategory: PropTypes.array,
};
