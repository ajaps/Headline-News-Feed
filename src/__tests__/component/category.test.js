import { shallow } from 'enzyme';
import React from 'react';
import ReactTestUtils from 'react-dom/test-utils';
import toJson from 'enzyme-to-json';
import Category from '../../components/Category.jsx';

jest.mock('../../api/NewsApi', () => ({
  fetchApiSources: () => Promise.resolve('getPromise')
}));

describe('Category component', () => {
  const categoryItem = { category: 'sport' };
  const currentCategory = 'sport';
  let app;
  beforeEach(() => {
    app = shallow(
      <Category
          category={categoryItem}
          currentCategory={currentCategory}
          key={categoryItem.category}
      />);
  });

  it('should render as expected', () => {
    const tree = toJson(app);
    expect(tree).toMatchSnapshot();
  });

  it('should call the set category function', () => {
    const renderedDoc = ReactTestUtils.renderIntoDocument(
      <Category
          category={categoryItem}
          currentCategory={currentCategory}
          key={categoryItem.category}
      />
    );
    const setCategoryBtn = renderedDoc.refs.category;
    ReactTestUtils.Simulate.click(setCategoryBtn);
  });
});
