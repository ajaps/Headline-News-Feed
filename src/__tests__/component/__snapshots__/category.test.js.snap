// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Category component should render as expected 1`] = `
<li
  className="active"
>
  <a
    onClick={[Function]}
  >
    sport
  </a>
</li>
`;
